core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```
@@ -712,37 +710,55 @@ private static void parseCopy(String field, ParseContext context) throws IOExcep
-            while (objectPath.length() > 0) {
-                String parentPath = "";
+            ObjectMapper[] mappers = new ObjectMapper[paths.length-1];
+            if (paths.length > 1) {
                 ObjectMapper parent = context.root();
-                posDot = objectPath.lastIndexOf('.');
-                if (posDot > 0) {
-                    parentPath = objectPath.substring(0, posDot);
-                    parent = context.docMapper().objectMappers().get(parentPath);
+                for (int i = 0; i < paths.length-1; i++) {
+                    context.path().add(paths[i]);
+                    mapper = context.docMapper().objectMappers().get(paths[i]);
+                    if (mapper == null) {
+                        // One mapping is missing, check if we are allowed to create a dynamic one.
+                        ObjectMapper.Dynamic dynamic = parent.dynamic();
+                        if (dynamic == null) {
+                            dynamic = dynamicOrDefault(context.root().dynamic());
+                        }
+                        if (dynamic == ObjectMapper.Dynamic.STRICT) {
+                            context.path().remove();
+                            throw new StrictDynamicMappingException(context.path().fullPathAsText(""), paths[i]);
+                        } else if (dynamic == ObjectMapper.Dynamic.TRUE) {
+                            context.path().remove();
+                            Mapper.Builder builder = context.root().findTemplateBuilder(context, paths[i], "object");
+                            if (builder == null) {
+                                builder = MapperBuilders.object(paths[i]).enabled(true).pathType(parent.pathType());
+                            }
+                            Mapper.BuilderContext builderContext = new Mapper.BuilderContext(context.indexSettings(), context.path());
+                            mapper = (ObjectMapper) builder.build(builderContext);
```

I think we should throw an exception if `mapper.nested() != Nested.NO`, as this would not make sense (and I'm pretty sure some users have dynamic templates to make all their objects nested by default).